Scipy optimize

michael-hartmann opened the pull request:
This PR uses the new optimize.minimize API also in the programs plot_gradient_descent.py and helper/compare_optimizers.py. This PR also completes the migration to the new API.

However, there is still a problem in helper/compare_optimizers.py. For the solver L-BFGS_B I get the warning
 ``OptimizeWarning: Unknown solver options: factr``.
According to the documentation, https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html, this option does exist. Unfortunately, I don't have time to investigate this issue myself.

If nobody has time to look at the problem right now, the best way would be to create a new issue, while closing the issues #215 and #300 .

GaelVaroquaux commented on advanced/mathematical_optimization/examples/helper/compare_optimizers.py:
```diff
@@ -15,30 +15,31 @@
-    "L-BFGS w f'":          my_partial(optimize.fmin_l_bfgs_b,
-                                        factr=10.0,
-                                        pgtol=1e-8, maxfun=1e7),
+    'Nelder-mead':          my_partial(method="Nelder-Mead", options = {
```
PEP8 tells us that there should no space around the equal sign after "options" (here and below)

GaelVaroquaux commented:
I think that the problem with factr is that it has been replaced by ftol. I would put "ftol=1e-13" and remove factr to avoid the warning.

Aside from this, and the PEP8 comment above, this is good to merge for me. Thanks a lot!

GaelVaroquaux commented:
@michael-hartmann : will you have time to address the comments, or should someone else try to finish off this PR?

gertingold commented:
@michael-hartmann is travelling and I am not sure that he will be able to react quickly. Should I finish his work?

GaelVaroquaux commented:
If you think that he will come back to this in a week or two, let's leave
it for him. I want to do a release, but it can wait a week or two.

If you have spare cycles, I think that improving the contributor guide
would be useful. At the sprint, it seemed to me that the workflow wasn't
obvious for newcomers.

gertingold commented:
It might be even a bit more than two weeks but there is no harm in waiting as long as it is ok for you. Completing the PR does not seem to be a big deal and it can be done relatively fast if necessary. If I find time, I will take a look at the contributor guide.

michael-hartmann commented:
I hope I find some time for it withing two weeks

michael-hartmann commented:
Sorry, I had some troubles while travelling and couldn't find time to make the changes earlier...

I have fixed the spaces in agreement with PEP8 and replaced factr by ftol.

GaelVaroquaux commented:
LGTM. Merging. Thank you!
